Snippets test infrastructure

IvanNovoselov opened the pull request:
Create snippets tests infrastructure.
### Details:
 - *Unit tests that cover common snippets transformations (e.g. insert_load_store and insert_movebroadcast)*
 - *Unit tests that cover device-specific snippets transformations (e.g. canonicalization and generation) for some dummy device*
 - *Functional tests with the ability to check the number of created subgraphs, and the total number of nodes in the exec graph*

### Tickets:
 - *74844*

tgubanova-lohika reviewed (changes requested):
Stop russian aggression in Ukraine!

IvanNovoselov commented:
@iefode @eshoguli, the PR is ready for review. Please, take a look

eshoguli reviewed (commented):
some comments were added, is still under review

iefode commented on src/tests/functional/inference_engine/CMakeLists.txt:
```diff
@@ -16,6 +17,7 @@ set(LINK_LIBRARIES
     openvino::conditional_compilation
     sharedTestClasses
     inference_engine_snippets
+    snippetsNgraphFunctions
```
Duplication

IvanNovoselov replied:
Removed inference_engine_snippets from the LINK_LIBRARIES

eshoguli commented on src/tests/functional/inference_engine/snippets/src/pass/lowering.cpp:
```diff
@@ -0,0 +1,80 @@
+    jitters[ngraph::snippets::op::Tile::get_type_info_static()] = dummy_functor;
+}
+
+void SnippetsLoweringTests::prepare() {
```
Not clear what `prepare` method do without implementation exploration. Can we remove `prepare()` and use `tokenize(function)` directly? `Tokenization` is term from snippets and `tokenize` method is clear.

iefode commented on src/tests/functional/inference_engine/snippets/src/pass/insert_load_store.cpp:
```diff
@@ -0,0 +1,74 @@
+// Copyright (C) 2022 Intel Corporation
```
Lets separate this tests to separated target because we try to test separated lib

IvanNovoselov replied:
Ok, let's do it, but after all other fixes are applied, since the new binary won't be compiled and run in CI (at least until this pr is merged).

iefode replied:
The PR with connected changes. I mean use the product config

IvanNovoselov replied:
Done, product config PR: https://github.com/intel-innersource/frameworks.ai.openvino.ci.product-configs/pull/1026

iefode commented on src/tests/functional/inference_engine/snippets/src/pass/insert_load_store.cpp:
```diff
@@ -0,0 +1,74 @@
+#include "common_test_utils/common_utils.hpp"
+#include <subgraph_lowered.hpp>
+
+namespace ngraph {
```
namespaxe ov {
namespace test {
namespace snippets {

IvanNovoselov replied:
Updated all the namespaces

iefode commented on src/tests/functional/shared_test_classes/src/base/snippets_test_utils.cpp:
```diff
@@ -0,0 +1,39 @@
+    size_t num_subgraphs = 0;
+    size_t num_nodes = 0;
+    for (const auto &op : compiled_model->get_ops()) {
+        auto layer_type = op->get_rt_info().at(ExecGraphInfoSerialization::LAYER_TYPE).as<std::string>();
```
we can use rtti instead of this construction

IvanNovoselov replied:
I can do it, no problem, but then we'll have to link sharedTestClasses against inference_engine_snippets, where the op::Subgraph is defined. Originally, I tried to avoid this dependence and used rt_info instead.

IvanNovoselov replied:
Decided to leave it as is to avoid additional linking dependencies.

iefode commented on src/tests/ie_test_utils/common_test_utils/snippets_test_utils.cpp:
```diff
@@ -0,0 +1,45 @@
+
+using ngraph::snippets::op::Subgraph;
+
+SnippetsRestoreResultInputName::SnippetsRestoreResultInputName() {
```
Why common? And why this namespace?>

ilya-lavrenov replied:
I agree, why this is common tests utils?

IvanNovoselov replied:
This is in common test utils because it's used in [cpu unit test](https://github.com/IvanNovoselov/openvino/blob/4767b226bea75a7b367067b4191eed87c42d9db8/src/tests/unit/cpu/ngraph_transformations/snipptes_mark_skipped.cpp#L23) and [ie funstional test](https://github.com/IvanNovoselov/openvino/blob/78548c806b880a820ae398308089f04fce24f28f/src/tests/functional/inference_engine/snippets/src/pass/collapse_subgraph.cpp#L19).

Please, also note that this transformation is needed only because TransformationTestsF calls ngraph::pass::CheckUniqueNames, which expect different Result-related ops and tensor naming rules compared to CPU plugin. See the explanation and proposed solutions in the [snippets_test_utils.hpp](https://github.com/IvanNovoselov/openvino/blob/4767b226bea75a7b367067b4191eed87c42d9db8/src/tests/ie_test_utils/common_test_utils/snippets_test_utils.hpp#L13-L24). So we won't need this transformation at all, if we resolve the conflict between ngraph::pass::CheckUniqueNames and CPU plugin Result-related nodes naming rules.

IvanNovoselov replied:
Another option is to include it inference_engine_snippets library. This transformation is rather small, so no significant binary size increase is expected. However, this pass is needed only for tests, so this is not the best solution.

IvanNovoselov replied:
As discussed with @dmitry-gorokhov offline, I'm going to update Subgraph node naming policy, so the subgraph will keep the name of the last attached node. This change will eliminate the need in this transformation.

Summary: The transformation is removed.

iefode commented on src/tests/ngraph_helpers/snippets_ngraph_functions/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+            PRIVATE
+                inference_engine_snippets
+        ADD_CPPLINT
+#        DEPENDENCIES
```
remove extra lines

IvanNovoselov replied:
Done, thanks for noticing

ilya-lavrenov commented on src/tests/functional/inference_engine/CMakeLists.txt:
```diff
@@ -26,6 +28,8 @@ set(DEPENDENCIES
     openvino_template_extension
     lptNgraphFunctions
+    inference_engine_snippets
+    snippetsNgraphFunctions
```
`DEPENDENCIES` should contain only dynamically loaded stuff, while common libraries are added as dependencies via `LINK_LIBRARIES` list

IvanNovoselov replied:
Valid point, removed.

eshoguli commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/snippets_helpers.hpp:
```diff
@@ -0,0 +1,81 @@
+/// \brief Base class for snippets-related subgraphs. Note that inputShapes size is model-specific
+/// and expected to be checked inside a child class constructor.
+/// \param inputShapes vector of input shapes accepted by the underlying model
+class SnippetsFunctionBase {
```
Why do we need base type for function builders? I think if we remove `SnippetsFunctionBase` then we will have more simpler source code.
For example: does base test type have pointer to the function builder instance to run virtual methods? - I didn't find it in existing source code.

IvanNovoselov replied:
As we discussed earlier https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r827767764, let's keep it for now.
I also provided [an example](https://github.com/IvanNovoselov/openvino/blob/aa40b2b7e5976560bfd7ae699957b69bca7ef30f/src/tests/functional/inference_engine/snippets/src/pass/insert_load_store.cpp#L36-L40) where a pointer to the base class is used, although I agree that it is not absolutely essential in this case, it still makes code more concise and readable.

IvanNovoselov replied:
I'm not sure we should do it because the lowering will be performed in SetUp in this case. We however need to test the lowering, so it should be performed inside TEST_P instead. Please see also this reply https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r830815416

eshoguli reviewed (commented):
General comments:
1. Please, put in order with file names and their content, for example: SnippetsMoveBroadcastTests is declared and implemented in insert_movebroadcast.h and insert_movebroadcast.cpp files.
2. Possible tests meaning is mixed: several functional tests (src\tests\functional\inference_engine\snippets), for example `SnippetsLoadStoreTests`, run tokenization before lowering. Can we run lowering only?
3. Type members usage to store intermediate values - please, simplify source code: there two comments for that (prepare(f.getOriginal()) and test parameters)

eshoguli commented:
please, squash commits before merge

IvanNovoselov commented on src/tests/unit/cpu/CMakeLists.txt:
```diff
@@ -12,8 +12,12 @@ addIeTargetTest(
+                $<TARGET_PROPERTY:openvino_intel_cpu_plugin,SOURCE_DIR>/src
+                $<TARGET_PROPERTY:openvino::conditional_compilation,INTERFACE_INCLUDE_DIRECTORIES>
+            PRIVATE
+                $<TARGET_PROPERTY:snippetsNgraphFunctions,SOURCE_DIR>/include
```
Agree, line 19 removed. Thanks!

ilya-lavrenov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+    UNITY
+    PCH PRIVATE "src/precomp.hpp"
+)
+openvino_developer_export_targets(COMPONENT tests TARGETS inference_engine_snippets)
```
good it, thanks

IvanNovoselov commented on src/tests/functional/shared_test_classes/include/shared_test_classes/base/snippets_test_utils.hpp:
```diff
@@ -0,0 +1,21 @@
+    void validateNumSubgraphs();
+    // Expected num nodes and subgraphs in exec graphs depends on the plugin
+    // pipeline, tokenization callback for example. Therefore, they have to be provided manually.
+    size_t ref_num_nodes = 0;
+    size_t ref_num_subgraphs = 0;
```
Absolutely agree, but lets implement originalLayerNames check in a separate PR.

IvanNovoselov commented on src/tests/functional/plugin/shared/src/snippets/add_convert.cpp:
```diff
@@ -0,0 +1,57 @@
+TEST_P(AddConvert, CompareWithRefImpl) {
+    run();
+    validateNumSubgraphs();
+}
```
This is done intentionally, so the AddConvert could be easily removed. The *Convert functions were added to bypass snippets_mark_skipped logics on skipping eltwises after inputs. As soon as we support all the precisions in snippets, this skipping logics (and also the *Convert tests) should be removed.

IvanNovoselov commented on src/tests/functional/shared_test_classes/src/base/snippets_test_utils.cpp:
```diff
@@ -0,0 +1,39 @@
+        num_subgraphs += layerType == "Subgraph";
+        num_nodes++;
+    }
+    ASSERT_EQ(ref_num_nodes, num_nodes)
+                                << "Compiled model contains invalid number of nodes.";
```
No problem, corrected.

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/src/pass/lowering.cpp:
```diff
@@ -0,0 +1,80 @@
+        m.run_passes(f);
+}
+
+void SnippetsLoweringTests::getSubgraph(std::shared_ptr<Model>& f) {
```
Here I wanted to avoid confusion with initOriginal() (also initReference and InitLowered) from SnippetsFunctionBase, because this function doesn't really initialize subgraph, it just finds already existing subgraph in the function.
If you think that this name is misleading I don't mind renaming, lets discuss an alternative offline.

IvanNovoselov replied:
Converted to static member as discussed in https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r825850184

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/src/pass/canonicalization.cpp:
```diff
@@ -0,0 +1,97 @@
+                                                                     {{1, 1,  2, 1}, {{1, 1, 2, 1, 1},  {0, 1, 2, 3, 1}, prec}},
+                                                                     {{1, 64, 1, 1}, {{1, 4, 1, 1, 16}, {0, 1, 2, 3, 1}, prec}}};
+
+INSTANTIATE_TEST_SUITE_P(BroadcastBlockedBlocked, SnippetsCanonicalizationTests,
```
1. Agree, will fix.
2. Agree in general, but it's not clear how to differentiate between different test suite instantiations. For example, SnippetsMoveBroadcastTests class covers two different cases BroadcastOnBothInputs and NoBroadcast [source](https://github.com/IvanNovoselov/openvino/blob/3b818e9f3592494a1f3ffbffda9b22b096842017/src/tests/functional/inference_engine/snippets/src/pass/insert_movebroadcast.cpp#L70-L80).
3. Also good point, but the test name won't be explanatory then. Imagine you have CanonicalizationTests failed in CI, but have never heard about snippets before (however, agree that one can always search for name in sources). 
Lets discuss 2 and 3 offline. We need a naming convention here.

IvanNovoselov replied:
Regarding 2:
Let's align test naming with CPU plugin: test ids are unique and formed using the pattern smoke_[layer name]\_[description], so it'll be smoke_Snippets_[description] in our case. Please [this test](https://github.com/openvinotoolkit/openvino/blob/a79830cb55623ce2654f0685b0ca51b3f412195a/src/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/batch_to_space.cpp#L99-L109) or [this test](https://github.com/openvinotoolkit/openvino/blob/a79830cb55623ce2654f0685b0ca51b3f412195a/src/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/bucketize.cpp#L55-L56) for example.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/snippets_helpers.hpp:
```diff
@@ -0,0 +1,81 @@
+/// \brief Base class for snippets-related subgraphs. Note that inputShapes size is model-specific
+/// and expected to be checked inside a child class constructor.
+/// \param inputShapes vector of input shapes accepted by the underlying model
+class SnippetsFunctionBase {
```
Current implementation guarantees that developer-provided functions (via overloading initOriginal and initLowered) are validated before any other transformations are performed. Thus we reduce the possibility that a test developer will provide an irrelevant function. Let's discuss whether we need this functionality.

IvanNovoselov replied:
After offline discussion with @eshoguli and then with @dmitry-gorokhov decided to keep the current structure for now. The reason is that the base class provides a unified interface (of cause at some additional cost, which is however negligible), and we might want to employ the introduced polymorphism in the future. If this is not the case, we can always easily get rid of the SnippetsFunctionBase.

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/src/pass/canonicalization.cpp:
```diff
@@ -0,0 +1,97 @@
+}
+
+TEST_P(SnippetsCanonicalizationTests, CompareWithRefImpl) {
+    const auto &f = AddFunction(input_shapes);
+    function = f.getOriginal();
+    function_ref = f.getReference();
+    prepare();
+    Shape canonical_output_shape = canonicalize(input_blocked_shapes, output_blocked_shapes);
```
Fixed, please see reply to https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r825837345

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/src/pass/insert_load_store.cpp:
```diff
@@ -0,0 +1,74 @@
+
+TEST_P(SnippetsLoadStoreTests, CompareWithRefEltwise) {
+    const auto &f = EltwiseFunctionThreeInputsLowered(input_shapes, broadcast_shapes);
+    function = f.getOriginal();
+    function_ref = f.getLowered();
+
+    prepare();
+    lower();
```
Agree, good point, thank you. Please also note that all ie functional snippets tests (and test utils) were refactored to address this comment.

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/include/pass/lowering.hpp:
```diff
@@ -0,0 +1,55 @@
+    void lower();
+
+private:
+    std::shared_ptr<Subgraph> subgraph;
```
Agree. This was addressed during the refactoring https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r827981674

IvanNovoselov commented:
@eshoguli, regarding your general comment https://github.com/openvinotoolkit/openvino/pull/10605#pullrequestreview-908631174:
1. Agree, fixed.
2. Because we need subgraph op to run lowering ([example](https://github.com/IvanNovoselov/openvino/blob/aa40b2b7e5976560bfd7ae699957b69bca7ef30f/src/tests/functional/inference_engine/snippets/src/lowering_utils.cpp#L58-L63)). Of course we can construct an appropriate sugraph op manually, but it's just much easier to tokenize automatically.
3. Agree, fixed. Please see also corresponding replies to your comments.

IvanNovoselov commented:
Had to force-push because of major namespace refactoring in CPU plugin.

IvanNovoselov commented:
There is still one unaddressed comment from @iefode (https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r825851093), but I suggest we fix it at the end of this review. For now, @eshoguli please take a second look.

eshoguli reviewed (approved):
It's OK to merge now, but let's discuss and fix comments later.
@iefode, can you, please, answer on questions.

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/src/pass/insert_load_store.cpp:
```diff
@@ -0,0 +1,72 @@
+TEST_P(InsertLoadStoreTests, ThreeInputsEltwise) {
+    auto subgraph = getLoweredSubgraph(snippets_function->getOriginal());
+    function = subgraph->get_body();
+    function_ref = snippets_function->getLowered();
```
Please, keep in mind that this is a lowering test, so the `function` should store a pointer to a lowered subgraph body. So we can initialize `function` in SetUp() only if we perform lowering there, but this is not an option because we actually testing the lowering.
The only thing we can do in regard of this comment is to initialize `function = snippets_function->getOriginal()`  in SetUp() and then reset it to `function = subgraph->get_body()` in TEST_P. However, I'm not sure it'll be clearer. What do you think?

iefode replied:
I am not sure about snippet functions and i need more details. But looks like what @eshoguli is right and this logic should be moved to `SetUp`

IvanNovoselov replied:
Decided to leave it as is, since no better solutions were suggested.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/subgraph_simple.hpp:
```diff
@@ -0,0 +1,133 @@
+    }
+protected:
+    std::shared_ptr<ov::Model> initOriginal() const override;
+//    std::shared_ptr<ov::Model> initReference() const override;
```
Removed, thanks for noticing.

IvanNovoselov commented on src/tests/functional/plugin/cpu/shared_tests_instances/snippets/conv_eltwise.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace {
+
+    const std::vector<InferenceEngine::Precision> netPrecisions = {
```
We have both [ov::element](https://github.com/openvinotoolkit/openvino/blob/f45991bd64fd47cb38898355557177d0a68461ab/src/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/eltwise.cpp#L44-L48) and [InferenceEngine::Precision](https://github.com/openvinotoolkit/openvino/blob/a79830cb55623ce2654f0685b0ca51b3f412195a/src/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/clamp.cpp#L30-L35) in test instantiations, so I don't which is preferable and why.

Anyway, we don't actually use precision parameter in these tests (since we support only fp32 for now), so I removed the unused parameter.

iefode replied:
Please align the test with OV2.0

IvanNovoselov replied:
Ok, Done.

IvanNovoselov commented on src/tests/functional/plugin/cpu/shared_tests_instances/snippets/conv_eltwise.cpp:
```diff
@@ -0,0 +1,38 @@
+                                 ::testing::Values(7), //num nodes = 7: Convert + Convolution + Subgraph + Reorders
+                                 ::testing::Values(1), // num subgraphs = 0: Mul (2 inputs) can't be fused into Conv => Subgraph is created
+                                 ::testing::Values(CommonTestUtils::DEVICE_CPU)),
+                         ConvEltwise::getTestCaseName);
```
The reason is to make it visible, what's included in ::testing::Combine(...). 
Ok, the indent is removed.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/subgraph_simple.hpp:
```diff
@@ -0,0 +1,133 @@
+//  HSwish   Log
+//      Multiply
+//       Result
+class EltwiseLogLoop : public SnippetsFunctionBase {
```
Renamed, thanks for noticing.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/subgraph_customizable.hpp:
```diff
@@ -0,0 +1,45 @@
+//           Relu
+//           Sqrt
+//          Result
+class ConvMulActivation : public SnippetsFunctionCustomizable {
```
Renamed, thanks for noticing.

IvanNovoselov replied:
Regarding https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r830525287 :
I think we shouldn't do it, because the lowering will be performed in SetUp in this case. We however need to test the lowering, so it should be performed inside TEST_P instead. Please see also thi reply https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r830815416

IvanNovoselov commented on src/tests/functional/inference_engine/snippets/include/lowering_utils.hpp:
```diff
@@ -0,0 +1,49 @@
+
+class LoweringTests : public TransformationTestsF {
+protected:
+    static std::shared_ptr<ngraph::snippets::op::Subgraph> getSubgraph(const std::shared_ptr<Model>& f);
```
We sure can, but let's do it in a separate PR, where such tests will be introduced.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/subgraph_simple.hpp:
```diff
@@ -0,0 +1,133 @@
+};
+/// EltwiseFunctionThreeInputs with Converts after inputs to to WA CPU-specific disabling after inputs
+/// See AddConvert for details.
+class EltwiseFunctionThreeInputsConvert : public SnippetsFunctionBase {
```
Ok, done

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/snippets_helpers.hpp:
```diff
@@ -0,0 +1,81 @@
+        throw std::runtime_error("initLowered() for this class is not implemented");
+    }
+
+    ov::element::Type_t precision;
+    std::vector<Shape> input_shapes;
```
Agree, fixed.

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/subgraph_lowered.hpp:
```diff
@@ -0,0 +1,54 @@
+class EltwiseFunctionThreeInputsLowered : public EltwiseFunctionThreeInputs {
+public:
+    explicit EltwiseFunctionThreeInputsLowered(std::vector<Shape> inputShapes, std::vector<Shape> broadcastShapes) :
+            EltwiseFunctionThreeInputs(std::move(inputShapes)), broadcast_shapes{std::move(broadcastShapes)} {
```
Updated clang-tidy guide ([modernize-pass-by-value] (https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html)) suggests to replace const references used to initialize class members with pass-by value + move. I tried to comply with these recommendations while writing the tests.

Anyway, I don't think that we will be able to detect any performance difference between the "traditional" const reference way and the updated modernize-pass-by-value, so I changed the function signatures back to the "traditional" way (although now we have some Clang-Tidy warnings in the code).

IvanNovoselov commented on src/tests/ngraph_helpers/snippets_ngraph_functions/include/snippets_helpers.hpp:
```diff
@@ -0,0 +1,81 @@
+public:
+    SnippetsFunctionBase() = delete;
+
+    explicit SnippetsFunctionBase(std::vector<Shape> &inputShapes, ov::element::Type_t precision = element::f32)
```
Fixed, please see https://github.com/openvinotoolkit/openvino/pull/10605#discussion_r830906942

IvanNovoselov commented:
@dmitry-gorokhov please, review

IvanNovoselov commented:
@iefode , @ilya-lavrenov please take a second look

IvanNovoselov commented on src/tests/functional/plugin/shared/include/snippets/add_convert.hpp:
```diff
@@ -1,34 +1,34 @@
 namespace LayerTestsDefinitions {
 
 typedef std::tuple<
         InferenceEngine::Precision,  // Network Precision
```
Ok, done

iefode reviewed (approved):
The changes is ok to me in general
